kramer/src/main/kotlin/kramer/MavenRepo.kt

```diff
@@ -0,0 +1,331 @@
+import com.github.ajalt.clikt.core.requireObject
+import com.github.ajalt.clikt.parameters.options.default
+import com.github.ajalt.clikt.parameters.options.defaultLazy
+import com.github.ajalt.clikt.parameters.options.option
+import com.github.ajalt.clikt.parameters.options.required
+import com.github.ajalt.clikt.parameters.types.int
+import com.github.ajalt.clikt.parameters.types.path
+import com.squareup.tools.maven.resolution.ArtifactResolver
+import com.squareup.tools.maven.resolution.FetchStatus
+import com.squareup.tools.maven.resolution.FetchStatus.RepositoryFetchStatus.SUCCESSFUL
+import com.squareup.tools.maven.resolution.FetchStatus.RepositoryFetchStatus.SUCCESSFUL.FOUND_IN_CACHE
+import com.squareup.tools.maven.resolution.Repositories
+import com.squareup.tools.maven.resolution.ResolvedArtifact
+import java.io.IOException
+import java.net.URI
+import java.nio.file.FileSystem
+import java.nio.file.FileSystemAlreadyExistsException
+import java.nio.file.FileSystems
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.Collections
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+import javax.xml.parsers.SAXParserFactory
+import kotlin.system.measureTimeMillis
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.filterNot
+import kotlinx.coroutines.flow.flatMapMerge
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.runBlocking
+import org.xml.sax.Attributes
+import org.xml.sax.SAXException
+import org.xml.sax.helpers.DefaultHandler
+
+class MavenRepo(fs: FileSystem = FileSystems.getDefault()) : CliktCommand(name = "gen-maven-repo") {
+  private val workspace: Path by option(
+    "--workspace",
+    help = "Path to the workspace to be generated."
+  )
+    .path(canBeFile = false, canBeSymlink = false, canBeDir = true)
+    .default(fs.getPath("${System.getProperties()["java.io.tmpdir"]}", "bazel/maven"))
+
+  /** Name of the workspace - nearly always the nearest directory name of the workspace path */
+  private val workspaceName: String by option("--workspace-name")
+    .defaultLazy { workspace.fileName.toString() }
+
+  private val configFile by option("--configuration").path(mustExist = true).required()
+
+  private val threadCount: Int by option("--threads").int().default(1)
+
+  internal val kontext by requireObject<Kontext>()
+
+  internal data class IndexEntry(
+    val dependants: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap()),
+    val versions: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap())
+  )
+
+  @FlowPreview
+  @ExperimentalCoroutinesApi
+  override fun run() {
```

Comment:
Please decompose this into some top level functions -- reasoning about the scope is not fun now, I can't imagine doing it 2 years when it needs maintenance.

There really shouldn't be a 200 line main method with multiple nested variable scopes.

Reply:
You've not done a lot of rx have you?

I'll do it. It's just that these lamdas are capturing lambdas, so they're crossing scope boundaries to handle the outer variables, and they're all going to have to go onto ever grosser function parameter lists. I expect it'll be less readable. But I'll give it a shot.

Reply:
As I said elsewhere, I was wrong, I restructured it all, the flow is now one-screen or less, and each step is either a function that returns a flow generator for that step (encapsulating the logic of the step) or a function that does the transformation for that step. (Only the flatMapMerge steps needed flow generators).

I was right that I needed to pipe in parameters from the outside context, and I had to make an AtomicInteger for the exit code since I needed a holder pattern that was thread-safe. It could slow things down in error cases, but only for the duration of an atomic integer set() operation, which is trivial.

So I think it's quite a bit easier to handle, and you were right to push. Made me consider a totally different angle (functions that returned the flow generator, rather than just what was in the flow). Also, your inter-flow result types which didn't make sense in-line made a lot more sense for separated functions.

Reply:
I have worked with RX -- I find it to be a bit of a solution looking for a problem.

I've got a long and weird JS background, and therefore opinions about cognitive overhead due to scope complexity. I find a great deal of benefit from streams, but I find the use of nested unnamed lambdas to be, well, in need of either massive commenting for each one or pulled to separate functions. The modern lambda revolution kicked off by Java 8 [citation needed] and popularized by Groovy, Kotlin, et. al. has given Guido a lot of weight for the single line lambda choice. 

Personally, I prefer to push towards Knuth's Literate programming. E.g. I want the language feature to get out of the way and let me see the intent. Or, as one of my old colleagues liked to quote "As simple as possible and no simpler" 

As for the concern about too many variables, it's usually and indication that the information is not grouped in an efficient manner. Programming is, in my understanding, an expression of who we think to solve a problem. Therefore, when working on code that will be maintained, one should strive for objective expressions of a solution -- as opposed to subjective. And let us note that "Object" is very much part of objective. :D